pymc3/distributions/multivariate.py

```diff
@@ -1091,3 +1091,183 @@ def logp(self, x):
                      eta > 0,
                      broadcast_conditions=False
         )
+
+
+class MatrixNormal(Continuous):
+    R"""
+    Matrix-valued normal log-likelihood.
+
+    ===============  =====================================
+    Support          :math:`x \in \mathbb{R}^{q \times p}`
+    Mean             :math:`\mu`
+    Right Variance   :math:`T_R^{-1}`
+    Left Variance    :math:`T_L^{-1}`
+    ===============  =====================================
+
+    Parameters
+    ----------
+    mu : array
+        Vector of means.
+    rcov : array
+        Right (or row) pxp covariance matrix. Defines variance within rows.
+        Exactly one of rcov or rchol is needed.
+    rchol : array
+        Cholesky decomposition of pxp covariance matrix. Defines variance
+        within rows. Exactly one of rcov or rchol is needed.
+    lcov : array
+        Left (or column) qxq covariance matrix. Defines variance within
+        columns. Exactly one of lcov or lchol is needed.
+    lchol : array
+        Cholesky decomposition of qxq covariance matrix. Defines variance
+        within columns. Exactly one of lcov or lchol is needed.
+
+    Examples
+    --------
+    ?
+    """
+
+    def __init__(self, mu=0, rcov=None, rchol=None, rtau=None,
+                 lcov=None, lchol=None, ltau=None, *args, **kwargs):
+
+        self.setup_matrices(rcov, rchol, rtau, lcov, lchol, ltau)
+
+        shape = kwargs.pop('shape', None)
+        assert len(shape) == 2, "only 2d tuple inputs work right now: qxp"
+        self.shape = shape
+
+        super(MatrixNormal, self).__init__(shape=shape, *args, **kwargs)
+
+        self.mu = tt.as_tensor_variable(mu)
+
+        self.mean = self.median = self.mode = self.mu
+
+        # self.solve_lower = tt.slinalg.Solve(A_structure="lower_triangular")
+        # self.solve_upper = tt.slinalg.Solve(A_structure="upper_triangular")
+        self.solve_lower = tt.slinalg.solve_lower_triangular
+        self.solve_upper = tt.slinalg.solve_upper_triangular
+
+    def setup_matrices(self, rcov, rchol, rtau, lcov, lchol, ltau):
+        # Step methods and advi do not catch LinAlgErrors at the
```

Comment:
NUTS should catch LinAlgErrors now. Maybe check how much faster we things get if we set `nofail=False`?